ytypes/gnmi.go

```diff
@@ -0,0 +1,150 @@
+package ytypes
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/openconfig/goyang/pkg/yang"
+	"github.com/openconfig/ygot/ygot"
+
+	gpb "github.com/openconfig/gnmi/proto/gnmi"
+)
+
+// UnmarshalNotification unmarshals a Notification on a root GoStruct specified by
+// "schema", and returns a reference to it.
+//
+// If preferShadowPath is specified, then the shadow path values are
+// unmarshalled instead of non-shadow path values when GoStructs are generated
+// with shadow paths.
+func UnmarshalNotification(schema *Schema, req *gpb.Notification, preferShadowPath bool) (ygot.GoStruct, error) {
+	return UnmarshalSetRequest(schema, &gpb.SetRequest{
+		Prefix: req.Prefix,
+		Delete: req.Delete,
+		Update: req.Update,
+	}, preferShadowPath)
+}
+
+// UnmarshalSetRequest applies a SetRequest on a root GoStruct specified by
+// "schema", and returns a reference to it.
+//
+// If preferShadowPath is specified, then the shadow path values are
+// unmarshalled instead of non-shadow path values when GoStructs are generated
+// with shadow paths.
+func UnmarshalSetRequest(schema *Schema, req *gpb.SetRequest, preferShadowPath bool) (ygot.GoStruct, error) {
+	root := schema.Root
+	node, nodeName, err := getOrCreateNode(schema.RootSchema(), root, req.Prefix, preferShadowPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// Process deletes, then replace, then updates.
+	if err := deletePaths(schema.SchemaTree[nodeName], node, req.Delete, preferShadowPath); err != nil {
+		return nil, err
+	}
+	if err := replacePaths(schema.SchemaTree[nodeName], node, req.Replace, preferShadowPath); err != nil {
+		return nil, err
+	}
+	if err := updatePaths(schema.SchemaTree[nodeName], node, req.Update, preferShadowPath); err != nil {
+		return nil, err
+	}
+
+	vroot, ok := root.(validatedGoStruct)
+	if !ok {
+		return nil, fmt.Errorf("schema root cannot be validated: (%T, %v)", root, root)
+	}
+	if err := vroot.ΛValidate(); err != nil {
+		return nil, fmt.Errorf("SetRequest is not schema-compliant: %v", err)
+	}
+
+	return vroot, nil
+}
+
+// validatedGoStruct is an interface used for validating GoStructs.
+// This interface is implemented by all Go structs (YANG container or lists),
+// regardless of generation flag.
+type validatedGoStruct interface {
```

Comment:
This is also an unexported type. I created this because I don't want to import the burden of ValidatedGoStruct, which has other functions that are unnecessary.

Reply:
Why is this a "burden" are there cases where you have types that don't implement the functions? It seems a little odd to me, because whilst this is a valid interface, it duplicates something that we do have today.